### ISSUE TITLE
5ttgen hsvs: Fix erroneous behaviour

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#2365]
Erroneous results observed in `5ttgen hsvs` outputs, but problem stems from a combination of `voxel2mesh` / `meshfilter smooth` / `mesh2voxel`. As [reported on forum](https://community.mrtrix.org/t/5ttgen-hsvs-failing-to-segment-regions/4498).

The [`mesh2image` algorithm relies on a sequence of steps](https://github.com/MRtrix3/mrtrix3/blob/3.0.3/src/surface/algo/mesh2image.cpp#L42-L556). Initially, it identifies the set of voxels intersected by the surface. It then attempts to classify the voxels adjacent to these intersected voxels as either presiding inside the surface or lying outside of it, based on the surface normals. It then looks for connected-component voxel clusters for which there is a consensus regarding this inside / outside classification (for robustness against having been provided with erroneous normal vectors, which can happen with certain relevant data sources). If this process doesn't work correctly, and all voxels outside the surface get erroneously classified as instead residing within the surface, everything downstream breaks.

One circumstance in which this can occur is if the Marching Cubes algorithm (i.e. `voxel2mesh`) is executed on a mask containing single voxel, followed by surface-based smoothing. 7b2b2152 attempted to mitigate this within #1386 by skipping the smoothing step for single-voxel mask images. However if a mask image contains **multiple isolated voxels**, that skip is not triggered, and so the same problem manifests.

So there are multiple possible avenues for rectification:

- The check introduced in 7b2b2152 should ideally first be running a connected-component analysis on the surface data (requested in #773), checking the sizes of the individual components, and then smoothing them individually (the current smoothing algorithm will not allow disconnected components to influence one another, the only difference is whether or not individual connected components would be *excluded* from smoothing);

- Further improvement to the `mesh2voxel` algorithm to make it more robust against structures that are sub-voxel in size (potentially with erroneous surface normals). 
I think what's specifically needed is to take an existing solution to another problem, being [counting how many of the eight corners of the image are in the connected component](https://github.com/MRtrix3/mrtrix3/blob/3.0.3/src/surface/algo/mesh2image.cpp#L299-L322), and override the surface-normal-consensus-based "internal" classification of the connected component if all eight corners of the image are in the connected component. This would technically break the unlikely case where all eight corners of the image are in fact inside of the surface but some voxels in the image are outside of the surface, but I'm not sure if it's worth trying to devise a solution that preserves that case.

**Comment by civier** [comment]
Hi @Lestropie

Not sure if it's related, but a student I co-supervise (Lizanne Arnoldy @lizannea) is running 5ttgen hsvs in MRtrix version 3.0.3, and it creates a Right_vessel.mif that is all 1s other than where the right vessel is. Unfortunately, there was no error message printed, so it has been quite labourious to understand what's going on. Is it related to the bug you describe above? After using the -nocleanup flag, we noticed that the mesh in the Right_vessel.vtk file was extremely small, and I provide it below for your reference. In the same participant, left vessel wasn't detected at all (vtk file has no polygons).

Best,
Oren




Right-vessel.vtk
============
# vtk DataFile Version 1.0

ASCII
DATASET POLYDATA
POINTS 12 float
22.63955212153407 30.97421935399937 -10.940566420457362
22.305940970424611 31.307830503862988 -10.940566420467334
22.639552121115798 31.307830504677927 -11.27417757115799
22.973163271806456 31.307830505511212 -10.940566420467865
21.639552000751614 31.974219471058245 -10.940566420486373
21.305940849642511 32.307830620921507 -10.940566420496083
21.639552000333609 32.307830621736443 -11.274177571187183
22.639552120697001 31.641441655374837 -10.940566420477845
21.973163151023993 32.307830622570073 -10.940566420496793
21.639551999914897 32.641441772433339 -10.940566420506503
22.639552121115269 31.307830504696277 -10.60695526977721
21.639552000332898 32.307830621755144 -10.606955269805693
POLYGONS 16 64
3 2 0 1
3 0 2 3
3 6 4 5
3 1 7 2
3 4 6 8
3 3 2 7
3 6 5 9
3 8 6 9
3 0 10 1
3 0 3 10
3 4 11 5
3 7 1 10
3 8 11 4
3 3 7 10
3 5 11 9
3 8 9 11

**Comment by Lestropie** [comment]
Yup, that'll be the same thing. Short term, if you modify the relevant aparc image replacing right hemisphere "vessel" voxels with just WM, `5ttgen hsvs` should then work. 

I don't have as much experience working with surface data as others in the field, so I don't know if there are preferable solutions to this problem.

It would help if I could have access to any cases where this problem arises so that I can see the relevant surfaces directly to confirm whether the smoothing filter is doing something detrimental to the surface normals, and test any proposed fix.
